pkg/sdn/plugin/sdn-cni-plugin/openshift-sdn.go

```diff
@@ -0,0 +1,463 @@
+	}
+
+	vnid := strconv.FormatUint(uint64(podInfo.Vnid), 10)
+
+	ingress, egress, err := getBandwidth(podInfo)
+	if err != nil {
+		return nil, "", "", "", false, err
+	}
+
+	macvlan, err := wantsMacvlan(podInfo)
+	if err != nil {
+		return nil, "", "", "", false, err
+	}
+
+	return nodeConfig, vnid, ingress, egress, macvlan, nil
+}
+
+// Returns host veth name, container veth MAC, and pod IP
+func getVethInfo(netns, containerIfname string) (netlink.Link, string, string, error) {
+	var (
+		peerIfindex int
+		contVeth    netlink.Link
+		err         error
+		podIP       string
+	)
+
+	containerNs, err := ns.GetNS(netns)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("Failed to get container netns: %v", err)
+	}
+	defer containerNs.Close()
+
+	err = containerNs.Do(func(ns.NetNS) error {
+		contVeth, err = netlink.LinkByName(containerIfname)
+		if err != nil {
+			return err
+		}
+		peerIfindex = contVeth.Attrs().ParentIndex
+
+		addrs, err := netlink.AddrList(contVeth, syscall.AF_INET)
+		if err != nil {
+			return fmt.Errorf("failed to get container IP addresses: %v", err)
+		}
+		if len(addrs) == 0 {
+			return fmt.Errorf("container had no addresses")
+		}
+		podIP = addrs[0].IP.String()
+
+		return nil
+	})
+	if err != nil {
+		return nil, "", "", fmt.Errorf("Failed to inspect container interface: %v", err)
+	}
+
+	hostVeth, err := netlink.LinkByIndex(peerIfindex)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("Failed to get host veth: %v", err)
+	}
+
+	return hostVeth, contVeth.Attrs().HardwareAddr.String(), podIP, nil
+}
+
+func addMacvlan(netns string) error {
+	var defIface netlink.Link
+	var err error
+
+	// Find interface with the default route
+	routes, _ := netlink.RouteList(nil, netlink.FAMILY_V4)
+	for _, r := range routes {
+		if r.Dst == nil {
+			defIface, err = netlink.LinkByIndex(r.LinkIndex)
+			if err != nil {
+				return fmt.Errorf("Failed to get default route interface: %v", err)
+			}
+		}
+	}
+	if defIface == nil {
+		return fmt.Errorf("Failed to find default route interface")
+	}
+
+	containerNs, err := ns.GetNS(netns)
+	if err != nil {
+		return fmt.Errorf("Failed to get container netns: %v", err)
+	}
+	defer containerNs.Close()
+
+	return containerNs.Do(func(ns.NetNS) error {
+		err := netlink.LinkAdd(&netlink.Macvlan{
+			LinkAttrs: netlink.LinkAttrs{
+				MTU:         defIface.Attrs().MTU,
+				Name:        "macvlan0",
+				ParentIndex: defIface.Attrs().Index,
+			},
+			Mode: netlink.MACVLAN_MODE_PRIVATE,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create macvlan interface: %v", err)
+		}
+		l, err := netlink.LinkByName("macvlan0")
+		if err != nil {
+			return fmt.Errorf("failed to create find macvlan interface: %v", err)
+		}
+		err = netlink.LinkSetUp(l)
+		if err != nil {
+			return fmt.Errorf("failed to set macvlan interface up: %v", err)
+		}
+		return nil
+	})
+}
+
+func getIPAMConfig(nodeConfig *NodeConfig) []byte {
+	_, ipnet, _ := net.ParseCIDR(nodeConfig.NodeNetwork)
```

Comment:
Won't this panic if the CIDR is invalid?  Is NodeConfig guaranteed to be valid?  Would be better to return an error here.